File: apps/api/app/schemas/statistics.py
from typing import List, Optional

from pydantic import BaseModel


class CardStateCounts(BaseModel):
    total: int
    new: int
    learning: int
    review: int
    mastered: int


class WordRef(BaseModel):
    headword: str
    slug: str
    lapses: Optional[int] = None
    interval_days: Optional[float] = None


class WeakCategory(BaseModel):
    slug: str
    name_en: str
    name_uz: str
    name_ru: str
    emoji: Optional[str] = None
    card_count: int
    lapses: int


class DayCount(BaseModel):
    day: str
    count: int


class StatisticsOut(BaseModel):
    cards: CardStateCounts
    total_reviews: int
    accuracy_all: float
    accuracy_mature: float
    time_spent_ms: int
    rating_breakdown: dict
    reviews_by_day: List[DayCount]
    forgotten: List[WordRef]
    mastered: List[WordRef]
    weak_categories: List[WeakCategory]

File: apps 2/web/lib/statistics.ts
import { apiFetch } from "@/lib/api";

export interface CardStateCounts {
  total: number;
  new: number;
  learning: number;
  review: number;
  mastered: number;
}

export interface WordRef {
  headword: string;
  slug: string;
  lapses: number | null;
  interval_days: number | null;
}

export interface WeakCategory {
  slug: string;
  name_en: string;
  name_uz: string;
  name_ru: string;
  emoji: string | null;
  card_count: number;
  lapses: number;
}

export interface DayCount {
  day: string;
  count: number;
}

export interface Statistics {
  cards: CardStateCounts;
  total_reviews: number;
  accuracy_all: number;
  accuracy_mature: number;
  time_spent_ms: number;
  rating_breakdown: Record<string, number>;
  reviews_by_day: DayCount[];
  forgotten: WordRef[];
  mastered: WordRef[];
  weak_categories: WeakCategory[];
}

export interface HeatmapDay {
  day: string;
  reviews_count: number;
  xp_earned: number;
  goal_reached: boolean;
}

export const statisticsApi = {
  statistics: () => apiFetch<Statistics>("/me/statistics", { auth: true }),
  heatmap: (days = 120) =>
    apiFetch<{ days: HeatmapDay[] }>(`/me/heatmap?days=${days}`, { auth: true }),
};
